wiki/Writing-tests-with-good-behavioural-coverage.md

```diff
@@ -0,0 +1,1265 @@
+
+```
+@Test
+fun testWithdraw_withSufficientBalance_updatesBalance() {
+    val output = ByteArrayOutputStream()
+    System.setOut(PrintStream(output))
+
+    account.withdraw("user", "password", 200.0)
+
+    assertThat(output.toString().trim()).isEqualTo("Withdrew 200.0. New balance is 800.0")
+    System.setOut(System.out)
+}
+```
+
+**b. To verify with receipt:**
+
+Ensures that when a receipt is requested, it includes the correct balance details of the withdrawal.
+
+```
+@Test
+fun testWithdraw_withReceipt_generatesReceipt() {
+    val output = ByteArrayOutputStream()
+    System.setOut(PrintStream(output))
+
+    account.withdraw("user", "password", 200.0, needReceipt = true)
+
+    assertThat(output.toString().trim()).contains("Receipt: Withdrew 200.0. Current balance: 800.0")
+    System.setOut(System.out)
+}
+```
+
+**c. To verify with passbook details:**
+
+Confirms balance statement with the passbook file being updated when a file is provided.
+
+```
+@Test
+fun testWithdraw_withPassbook_updatesPassbook() {
+    val passbookFile = File("passbook.pdf")
+    passbookFile.createNewFile()
+    val output = ByteArrayOutputStream()
+    System.setOut(PrintStream(output))
+
+    account.withdraw("user", "password", 200.0, file = passbookFile)
+
+    // Read the passbook file and check its contents
+    val fileContents = passbookFile.readText()
+    assertThat(fileContents).contains("Withdrew 200.0. New balance is 800.0")
+
+    System.setOut(System.out)
+}
+```
+
+**d. To verify based on the log message:**
+
+Validates that the correct message about the withdrawal is logged.
+
+```
+@Test
+fun testWithdraw_withSufficientBalance_logsCorrectMessage() {
+    val output = ByteArrayOutputStream()
+    System.setOut(PrintStream(output))
+
+    account.withdraw("user", "password", 200.0)
+
+    assertThat(output.toString().trim()).contains("New balance is 800.0")
+    System.setOut(System.out)
+}
+```
+
+**e. To verify with the balance is updated correctly:**
+
+Ensures that the balance is updated correctly after a withdrawal.
+
+```
+@Test
+fun testWithdraw_withSufficientBalance_updatesBalanceCorrectly() {
+    account.withdraw("user", "password", 200.0)
+    assertThat(account.balance).isEqualTo(800.0)
+}
+```
+
+These tests cover various aspects of the withdrawal functionality, ensuring that the balance updates correctly with receipts, passbooks and output messages. Although the core functionality of withdrawing funds and updating the balance is consistent, it can be observed in multiple ways. Each test focuses on a specific verification method while ultimately validating the same core functionality.
+
+## Testing Practices for Effective Validation
+
+The following best practices are meant to assist in the organization and execution of tests. By following these guidelines, you can ensure that your testing process is efficient, thorough, and well-structured.
+
+### 1. Use of Helper Functions
```

Comment:
Important detail to include here (and maybe as part of the setup/teardown parts): **only** details not directly relevant to the _behavior_ of a test should be moved to helpers or the before/after functions. These aren't details that are necessary for the test to succeed (I assume all lines are necessary, otherwise they shouldn't be included). Instead, these are details that, if missing from the main test body, would make it confusing as a reader to verify that the implementation of a test actually matches the behavior it's trying to verify.

This is sort of the art form of writing tests: the more unrelated details that can be extracted, the better since it helps keep the test much more focused.

It's maybe also important to clarify somewhere that we use tests to verify implementation code is correct, but humans verify that test code is correct. For this reason, we err on the side of code duplication in tests (unlike production code) or other forms of code repetition since clarity is more important than conciseness in test writing.

Reply:
Understood, but the two approaches seem contradictory to how I had them earlier, (making it not so fitting in the after before annotations) so restructured it to be

### Structuring Test Functionalities 
- When and How to Divide Responsibilities
  - Helper Functions
  - @Before and @After Annotations
- When Not to Divide Responsibilities
  - Prioritizing Clarity Over Conciseness in Tests
- Importance of Descriptive Test Names

https://github.com/oppia/oppia-android/blob/1c1531d9d938dcee6c47078a1349d2ef2762de27/wiki/Writing-tests-with-good-behavioural-coverage.md#structuring-test-functionalities

will revise as per feedback.